### PR TITLE
Pr create floating ip on demand

### LINKED ISSUE #427: create floating ip on the fly if it's not created yet
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
per discussion on #422 ,we need a floating ip to be created first, this works but can be enhanced, so we can create a floating ip by using the 'master ip' in cluster actuator

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

### PR DESCRIPTION
**Pull request by sbueringer**
This PR creates machine floating ips if they don't exist. Depending on the interpretation fixes https://github.com/kubernetes-sigs/cluster-api-provider-openstack/issues/427 :)

**What this PR does / why we need it**:
Floating IP handling is now consistent. LoadBalancer & machine floating ip's are both created if they don't already exist. If they already exist they are just associated.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #427

### CODE REVIEW

**Comment by sbueringer** [comment]
/assign @jichenjc 

(merge after multi-node PR, but might help to verify the PR https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/421 on environments where floating ip cannot be created)
/hold

**Comment by sbueringer** [comment]
@jichenjc I rebases this PR. Diff looks now friendlier :)

Can you take a look? This should allow reusing existing floating ips for the LoadBalancer.

File: pkg/cloud/openstack/services/networking/floatingip.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	"fmt"
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/layer3/floatingips"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+	providerv1 "sigs.k8s.io/cluster-api-provider-openstack/pkg/apis/openstackproviderconfig/v1alpha1"
+	"time"
+)
+
+func (s *Service) CreateFloatingIP(clusterProviderSpec *providerv1.OpenstackClusterProviderSpec, ip string) error {
```
Comment (jichenjc) [inline pkg/cloud/openstack/services/networking/floatingip.go]: suggest we rename this to GetorCreateFloatingIP as this might just return if it exist

File: pkg/cloud/openstack/machine/actuator.go [hunk 2]
```diff
@@ -149,7 +155,13 @@ func (a *Actuator) Create(ctx context.Context, cluster *clusterv1.Cluster, machi
 	}
 
 	if machineProviderSpec.FloatingIP != "" {
-		err := computeService.AssociateFloatingIP(instance.ID, machineProviderSpec.FloatingIP)
+		err := networkingService.CreateFloatingIP(clusterProviderSpec, machineProviderSpec.FloatingIP)
```
Comment (jichenjc) [inline pkg/cloud/openstack/machine/actuator.go]: my env has some issue here
rule:create_floatingip and rule:create_floatingip:floating_ip_address) is disallowed by policy"
so I can't test, but code looks fine

File: pkg/cloud/openstack/services/networking/floatingip.go (hunk 1 above)
Comment (sbueringer) [reply pkg/cloud/openstack/services/networking/floatingip.go]: Yup

File: pkg/cloud/openstack/machine/actuator.go (hunk 2 above)
Comment (sbueringer) [reply pkg/cloud/openstack/machine/actuator.go]: But it works if the floating ip is already there before?

Comment (jichenjc) [reply pkg/cloud/openstack/machine/actuator.go]: yes, you are right, I tested that path ,so I think we are ready to go now

Comment (sbueringer) [reply pkg/cloud/openstack/machine/actuator.go]: Perfect. Didn't want to break your use case with that :)

File: pkg/cloud/openstack/services/userdata/kubeadm.go [hunk 3]
```diff
@@ -42,11 +42,14 @@ func generateKubeadmConfig(isControlPlane bool, bootstrapToken string, cluster *
 			// Set default values. If they are not set, these two properties are added with an
 			// empty string and the CoreOS ignition postprocesser fails with "could not find expected key"
 			if clusterConfigurationCopy.CertificatesDir == "" {
-				clusterConfigurationCopy.CertificatesDir = "/etc/kubernetes/pki"
+				clusterConfigurationCopy.CertificatesDir = kubeadmv1beta1.DefaultCertificatesDir
```
Comment (jichenjc) [inline pkg/cloud/openstack/services/userdata/kubeadm.go]: seems this is not related to your current PR...

Comment (sbueringer) [reply pkg/cloud/openstack/services/userdata/kubeadm.go]: Oh yeah. I'll move that to another one

Comment (sbueringer) [reply pkg/cloud/openstack/services/userdata/kubeadm.go]: Moved to here: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/444

**Comment by sbueringer** [comment]
/hold

Waiting until the new v1alpha2 controllers are merged

**Comment by sbueringer** [comment]
/hold cancel

I think it's safe to merge. I just have to make sure to not loose it in the controllers PR :)
